FEATURE: Add user option to toggle timezone adjustment

jancernik opened the pull request:
- The button only appears when `enable_timezone_offset_for_calendar_events` is set to true.
- The option is not persistent; it resets each time the calendar is loaded.
- By default, the offset is disabled and is configurable via `default_timezone_offset_user_option`.
- Similar to changing the timezone, it takes some time to rerender the events, depending on the number of them.

![demo](https://github.com/discourse/discourse-calendar/assets/66427541/e29ee987-c47c-4e49-8b1a-36d0abef423b)

pmusaraj commented on assets/javascripts/discourse/initializers/discourse-calendar.js:
```diff
@@ -295,11 +300,21 @@ function initializeDiscourseCalendar(api) {
+      customButtons:
+        siteSettings.enable_timezone_offset_for_calendar_events && {
+          timezoneOffset: {
+            text: "Timezone Offset",
```
Can we use the `globe` icon instead of a text label? It would take less space in the screen. You can try with either `iconNode("globe")` or `iconHTML("globel")` helpers.

jancernik replied:
The default icon support with the FullCalendar library is very limited. It doesn't allow custom icons or HTML. We would need to render it and then replace the innerHTML afterwards. But yes, I agree, an icon would be much better

pmusaraj replied:
Can we use our own button, outside the calendar, next to the timezone dropdown?

pmusaraj commented on assets/javascripts/discourse/initializers/discourse-calendar.js:
```diff
@@ -30,6 +33,8 @@ function initializeDiscourseCalendar(api) {
     return;
   }
 
+  let enableTimezoneOffset = siteSettings.default_timezone_offset_user_option;
```
Do we have too many site settings related to this feature now? Maybe worth a followup PR.

jancernik replied:
I wasn't thinking of adding this one, but since it takes a while to rerender the events when clicking the button, it makes it more reliable for testing. Maybe aother way could be to await for the the button state or something.
Yes, I'll probably have a followup PR revising all of this

pmusaraj commented on assets/javascripts/discourse/initializers/discourse-calendar.js:
```diff
@@ -20,7 +20,10 @@ function loadFullCalendar() {
 }
 
+let calendarPost;
+let calendar;
```
I see a lot of removed `calendar` references. Do we still need to declare this here?

jancernik replied:
Yes, since I had to access the `calendar` object from a function defined within its creation, I had to make the variable global. That's why I removed many of the references when passing it as an attribute between functions, but we still need it

pmusaraj reviewed (approved):
Looks good, mostly cosmetic comments.

jancernik commented:
I changed the code to manually add a button in the header instead. I used `iconHTML("globe")` instead of text.
It's nice because it required less refactoring, so I reverted those changes back.
I updated the gif in the description.

pmusaraj commented on assets/javascripts/discourse/initializers/discourse-calendar.js:
```diff
@@ -810,6 +827,38 @@ function initializeDiscourseCalendar(api) {
+    );
+    const timezoneButton = document.createElement("button");
+
+    timezoneButton.classList.add("fc-timezoneOffset-button");
+    timezoneButton.classList.add("fc-button");
```
```suggestion
    timezoneButton.classList.add("fc-timezoneOffset-button", "btn", "btn-icon");
```

I'd recommend using the button classes from core ("btn" and "btn-default") here. It makes the button styling more consistent, and you shouldn't need to add an active state class (below) or specific SCSS styling, it will inherit from core. 

Not sure this works 100%. Make sure you also look at the styleguide, `http://localhost:4200/styleguide/atoms/buttons`, it has great examples of button types and button states.

pmusaraj replied:
Also, for accessibility, we should also add a `title` or an `ariaLabel` property to the button.

pmusaraj reviewed (approved):
Looking good! Approving, but I do think the button CSS classes need a little bit of cleaning up to match core.

pmusaraj commented on assets/javascripts/discourse/initializers/discourse-calendar.js:
```diff
@@ -833,10 +833,11 @@ function initializeDiscourseCalendar(api) {
     );
     const timezoneButton = document.createElement("button");
 
+    timezoneButton.title = "Toggle timezone offset";
```
This needs to be localizable, something like: 

```javascript
I18n.t("discourse_calendar.toggle_timezone_offset_title");
```

and then `toggle_timezone_offset_title` should be added to client.en.yml.

pmusaraj reviewed (approved):
Sweet, LGTM!
